clients/cobol-lsp-vscode-extension/src/services/ControlFlowService.ts

```diff
@@ -103,75 +127,161 @@ class AnalysisTask {
 
 export class ControlFlowAnalysisService implements AnalysisServiceDelegate {
   private tasks: Map<string, AnalysisTask>;
-  private callbacks: Map<string, ControlFlowAnalysisCallback>;
+  private latestResults: Map<string, LatestResultData>;
   private diagnosticService: DiagnosticService;
+  private requestVersion: number = 0;
 
   public constructor(
     private mainChannel?: vscode.OutputChannel,
     private logChannel?: vscode.LogOutputChannel,
   ) {
     this.tasks = new Map<string, AnalysisTask>();
-    this.callbacks = new Map<string, ControlFlowAnalysisCallback>();
     this.diagnosticService = new DiagnosticService();
+    this.latestResults = new Map<string, LatestResultData>();
+    this.requestVersion = 1;
+  }
+
+  public async invalidate(documentUri: string, rejectPromise: boolean) {
+    this.logChannel?.debug(`Invalidate document: ${documentUri}`);
+
+    this.invalidatePromise(documentUri, rejectPromise);
+    await this.removeTask(documentUri);
+  }
+
+  public async getAnalysis(documentUri: string): Promise<AnalysisResult> {
+    this.logChannel?.debug(
+      `Get analysis request for the document: ${documentUri}`,
+    );
+
+    const latestResult = this.latestResults.get(documentUri);
+    if (latestResult) {
+      return latestResult.promise;
+    } else {
+      return this.createLatestResultPromise(documentUri, 0);
+    }
+  }
+
+  public makeControlFlowAstNotificationHandler() {
+    return (result: ApiResult) => {
+      this.handleControlFlowAst(result).catch(() => {});
+    };
   }
 
-  public queueAnalysis(programs: Program[], documentUri: string) {
+  async handleControlFlowAst(result: ApiResult) {
+    this.logChannel?.debug("Handle AST from backend");
+    if (result.documentUri) {
+      this.invalidatePromise(result.documentUri, false);
+      if (result.controlFlowAST.length > 0) {
+        this.queueAnalysis(result.controlFlowAST, result.documentUri);
+      }
+      await this.removeTask(result.documentUri);
```

Comment:
The removal should still be started before the next action is queued.
```suggestion
      const removal = this.removeTask(result.documentUri);
      if (result.controlFlowAST.length > 0) {
        this.queueAnalysis(result.controlFlowAST, result.documentUri);
      }
      await removal;
```